backend/public-graph/graph/resolver.go

```diff
@@ -372,7 +372,28 @@ func (r *Resolver) getMappedStackTraceString(ctx context.Context, stackTrace []*
 	return newMappedStackTraceString, mappedStackTrace, nil
 }
 
-func (r *Resolver) GetOrCreateErrorGroup(ctx context.Context, errorObj *model.ErrorObject, match *int) (*model.ErrorGroup, error) {
+func (r *Resolver) GetOrCreateErrorGroup(ctx context.Context, errorObj *model.ErrorObject, matchFn func() (*int, error)) (*model.ErrorGroup, error) {
+	match, err := matchFn()
+	if err != nil {
+		return nil, err
+	}
+	// no match means we are planning to create an error group
+	// we should lock the write path to make sure we do not write duplicate error groups
+	if match == nil {
+		key := fmt.Sprintf("GetOrCreateErrorGroup-project-%d", errorObj.ProjectID)
+		if mutex, err := r.Redis.AcquireLock(ctx, key, time.Minute); err == nil {
```

Comment:
if it isn't able to acquire the lock within the timeout, we'd expect duplicate error groups?

Reply:
yeah, should we return an error instead so that we can try again?